scripts/toolchain/install-toolchain.sh

```diff
@@ -0,0 +1,176 @@
+#!/bin/bash
+
+# prevent sourcing of this script, only allow execution
+$(return >/dev/null 2>&1)
+test "$?" -eq "0" && { echo "This script should only be executed." >&2; exit 1; }
+
+# exit on errors, undefined variables, ensure errors in pipes are not hidden
+set -Eeuo pipefail
+
+# set log id and use shared log function for readable logs
+declare mydir
+mydir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &>/dev/null && pwd -P)
+
+# Installs all toolchain utilities that are required to build hoprnet monorepo, including
+# - Node.js
+# - Yarn
+# - Typescript + related utilities, such as ts-node
+# - Rust (rustc, cargo)
+# - wasm-pack + wasm-opt, necessary to build WebAssembly modules
+# 
+# Currently tested for x86_64 and Alpine Linux based environments
+
+# @TODO adapt this script for macOS arm64 machines
+
+function usage() {
+  msg
+  msg "Usage: $0 [-h|--help] [--runtime-only]"
+  msg
+  msg "This script installs all required toolchain utilities to build hoprnet monorepo"
+  msg
+  msg "Use --runtime-only to install only those utilities that are necessary at runtime"
+  msg
+}
+
+declare install_all
+install_all="true"
+
+while (( "$#" )); do
+  case "$1" in
+    -h|--help)
+      # return early with help info when requested
+      usage
+      exit 0
+      ;;
+    --runtime-only)
+      install_all="false"
+      shift
+      ;;
+    -*|--*=)
+      usage
+      exit 1
+      ;;
+    *)
+      shift
+      ;;
+  esac
+done
+
+if [[ "${PATH}" =~ "/usr/local/bin" ]]; then
+    echo "Cannot install utilities. \"/usr/local/bin\" is not part of home-path."
+fi
+
+if ! [ -w "/usr/local/bin" ]; then
+    echo "Cannot install utilities. \"/usr/local/bin\" is not writable."
+fi
+
+if ! [ -d "/opt" ] && ! [ -w "/opt" ] || ! [ -w "/" ]; then
+    echo "Cannot install utilities. \"/opt\" does not exist or is not writable."
+fi
+
+declare download_dir="/tmp/hopr-toolchain/download"
+mkdir -p ${download_dir}
+
+function install_rustup() {
+    if ! command -v rustup; then
+        echo "Installing Rustup"
+        # Get rustup but install toolchain later
+        curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --default-toolchain none -y
+    fi
+}
+
+function install_cargo() {
+    if ! command -v cargo; then
+        declare rust_toolchain_toml_path="${mydir}/../../rust-toolchain.toml"
+        declare rust_version=$(sed -En 's/^channel = "([0-9a-z.]*)"$/\1/p' ${rust_toolchain_toml_path})
+        echo "Installing Cargo, Rust compiler v${rust_version}"
+        declare target=$(sed -En 's/^targets = \[ "([a-z0-9-]*)" \]$/\1/p' ${rust_toolchain_toml_path})
+        declare profile=$(sed -En 's/^profile = "([a-z]*)"$/\1/p' ${rust_toolchain_toml_path})
+
+        # Always install the version specified in `rust-toolchain.toml`
+        $HOME/.cargo/bin/rustup toolchain install ${rust_version} --profile ${profile} --target ${target}
+        source "$HOME/.cargo/env"
+    fi
+}
+
+function install_wasm_pack() {
```

Comment:
This function will only work for x86_64 architectures. Would be good to add a check at the beginning of the script to check if the architecture is the expected, to avoid running this script in other architectures at the moment?

Reply:
I've created the initial version for `x86_64` and left further adaptations out of scope. It's planned to have this also for `arm64`